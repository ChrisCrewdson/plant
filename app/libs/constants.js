module.exports = {
  uuidRE: /^[0-9a-f]{32}$/i,
  // Mongo _id regex: 57acfcf5ffaf0524572d58e8
  mongoIdRE: /^[0-9a-f]{24}$/i,
  maxImageFilesPerUpload: 100,
  awsBucketName: 'i.plaaant.com',
  imageSizeNames: ['orig', 'xl', 'lg', 'md', 'sm', 'thumb'],
  gisMultiplier: 10 ** 7,

  // Any user that is a member of a location should have one of these
  // roles for that location. The roles are, from left to right, most
  // permissive to least permissive.
  roles: ['owner', 'manager', 'member'],

  SERVICE_NAME: 'plant',
};
