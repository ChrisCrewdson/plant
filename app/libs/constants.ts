import { ImageSizeName } from '../../lib/db/mongo/model-note';

// A file for constants used on both the client and server

export const imageSizeNames: ImageSizeName[] = ['orig', 'xl', 'lg', 'md', 'sm', 'thumb'];

// Any user that is a member of a location should have one of these
// roles for that location. The roles are, from left to right, most
// permissive to least permissive.
export const roles: Role[] = ['owner', 'manager', 'member'];

export const uuidRE = /^[0-9a-f]{32}$/i;
export const mongoIdRE = /^[0-9a-f]{24}$/i; // Mongo _id regex 57acfcf5ffaf0524572d58e8
export const maxImageFilesPerUpload = 100;
export const awsBucketName = 'i.plaaant.com';

export const gisMultiplier = 10 ** 7;
export const SERVICE_NAME = 'plant';
