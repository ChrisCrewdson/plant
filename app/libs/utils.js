const constants = require('./constants');
const slug = require('slug');
const isDate = require('lodash/isDate');
const moment = require('moment');
const Immutable = require('immutable');

const {gisMultiplier} = constants;

// bson is currently not being explicitly installed in the project because
// mongodb depends on mongodb-core which depends on bson. The Npm 3 installer
// will therefore install bson as a top level dependency in node_modules. If
// this pattern is changed then we would need to install npm independently.
// Requiring only bson here achieves 2 things:
// 1. Fixes a problem that Webpack has when bundling this module and chaining
//    from mongodb down to bson and,
// 2. Reduces the size of the bundle that gets generated for the browser.
const bson = require('bson');
const {ObjectID} = bson;

function makeMongoId() {
  return new ObjectID().toString();
}

function makeSlug(text) {
  if(!text) {
    console.warn('text is falsey in makeSlug:', text);
    return '';
  }

  text = text.toString();
  text = text.replace(/\//g, ' ');
  return slug(text.toString(), {
    // replacement: '-',
    // symbols: true,
    // remove: /[.]/g,
    lower: true,
    // charmap: slug.charmap,
    // multicharmap: slug.multicharmap
  });
}

function makeUrl(first, user) {
  const userName = user.get('name');
  const _id = user.get('_id');

  return `/${first}/${makeSlug(userName)}/${_id}`;
}

/**
 * Make a /plants/user-name-slug/id url from user object
 * @param {Immutable.Map} user - an Immutable.js Map
 * @returns {string} - a url
 */
function makePlantsUrl(user) {
  return makeUrl('plants', user);
}

function makeLayoutUrl(user) {
  return makeUrl('layout', user);
}

/**
 * Convert a date like object to an Integer
 * @param {object} date - could be object, string or Integer
 * @returns {Integer} - a date in the form YYYYMMDD
 */
function dateToInt(date) {
  if(moment.isMoment(date)) {
    return dateToInt(date.toDate());
  } else if(isDate(date)) {
    return date.getFullYear() * 10000 +
      (date.getMonth() + 1) * 100 +
      date.getDate();
  } else if (typeof date === 'string') {
    return dateToInt(new Date(date));
  } else if (typeof date === 'number') {
    return date;
  } else {
    console.error('Unable to convert in dateToInt:', date);
    throw new Error(`dateToInt(${date})`);
  }
}

function intToDate(date) {
  const year = Math.round(date / 10000);
  const month = Math.round( (date - year * 10000) / 100);
  const day = Math.round( (date - (year * 10000 + month * 100)));
  return new Date(year, month - 1, day);
}

function intToMoment(date) {
  return moment(intToDate(date));
}

function intToString(date) {
  return intToMoment(date).format('MM/DD/YYYY');
}

/**
 * Converts the body of a POST/PUT to a plant object.
 * @param {object} body - POST/PUT body
 * @returns {object} - body with relavant fields converted to correct data type
 */
function plantFromBody(body) {
  const dateFields = ['plantedDate', 'purchasedDate', 'terminatedDate'];
  dateFields.forEach(dateField => {
    if(body[dateField]) {
      body[dateField] = parseInt(body[dateField], 10);
    }
  });
  if(typeof body.isTerminated === 'string') {
    body.isTerminated = body.isTerminated === 'true';
  }
  return body;
}

/**
 * Filters the plantIds array based on filter
 * @param {array} plantIds - original plantIds to filter
 * @param {Immutable.Map} plants - all the plants available to sort
 * @param {string} filter - optional text to filter title of plant
 * @returns {array} - an array of filtered plantIds
 */
function filterPlants(plantIds, plants, filter) {
  return filter
    ? plantIds.filter(plantId => {
      const plant = plants.get(plantId);
      return !plant || (plant.get('title') || '').toLowerCase().indexOf(filter) >= 0;
    })
    : plantIds;
}

/**
 * Sorts the plantIds based on the plant's title
 * @param {array} plantIds - original plantIds to filter
 * @param {Immutable.Map} plants - all the plants available to sort
 * @returns {array} - an array of sorted plantIds
 */
function sortPlants(plantIds, plants) {
  return plantIds.sort((a, b) => {
    const plantA = plants.get(a);
    const plantB = plants.get(b);
    if(plantA && plantB) {
      if(plantA.get('title') === plantB.get('title')) {
        return 0;
      }
      return plantA.get('title') > plantB.get('title') ? 1 : -1;
    } else {
      return 0;
    }
  });
}

/**
 * Filters the plantIds array and sorts based on the plant's title
 * @param {array} plantIds - original plantIds to filter
 * @param {Immutable.Map} plants - all the plants available to sort
 * @param {string} filter - optional text to filter title of plant
 * @returns {array} - an array of sorted and filtered plantIds
 */
function filterSortPlants(plantIds, plants, filter) {
  const filteredPlantIds = filterPlants(plantIds, plants, filter);

  return sortPlants(filteredPlantIds, plants);
}

/**
 * The values of the errors object are arrays. Take the first item out of each array.
 * @param {object} errors - values are arrays
 * @returns {object} - first element of value for each key
 */
function transformErrors(errors) {
  if(!errors) {
    return errors;
  }
  return Object.keys(errors).reduce((acc, key) => {
    acc[key] = errors[key][0];
    return acc;
  }, {});
}

function hasGeo() {
  return !!(navigator && navigator.geolocation);
}

function getGeo(options, cb) {
  if(!hasGeo()) {
    return cb('This device does not have geolcation available');
  }

  options = Object.assign({}, {
    enableHighAccuracy: true,
    timeout: 30000, // 10 seconds
  }, options);

  navigator.geolocation.getCurrentPosition(position => {
    // { type: "Point", coordinates: [ 40, 5 ] }
    // postion: {coords: {latitude: 11.1, longitude: 22.2}}
    const geoJson = {
      type: 'Point',
      coordinates: [
        position.coords.longitude,
        position.coords.latitude,
      ]
    };
    return cb(null, geoJson);
  }, positionError => {
    console.error('geolcation error:', positionError);
    return cb('There was an error get the geo position', positionError);
  }, options);
}


/**
 * Because math in JS is not precise we need to use integers
 * to subtract 2 number for GIS rebasing
 * @param {number} left - the left number in the operation
 * @param {number} right - the right number in the operation
 * @returns {number} - left - right
 */
function subtractGis(left, right) {
  // 7 decimal places in long/lat will get us down to 11mm which
  // is good for surveying which is what we're basically doing here
  return Math.round(left * gisMultiplier - right * gisMultiplier) / gisMultiplier;
}

/**
 * Rebase the long and lat of the plant locations
 * @param {array} plants - an array of plants with just the _id and loc fields
 * @returns {array} - same array of plants with the locations rebased to 0,0
 */
function rebaseLocations(plants) {
  if(!plants || !plants.length) {
    return plants;
  }

  const northWestPoints = plants.reduce((acc, plant) => {
    const [long, lat] = plant.loc.coordinates;
    acc.long = Math.min(acc.long, long);
    acc.lat = Math.min(acc.lat, lat);
    return acc;
  }, {long: 180, lat: 90});

  return plants.map(plant => {
    return {
      _id: plant._id.toString(),
      loc: {
        coordinates: [
          subtractGis(plant.loc.coordinates[0], northWestPoints.long),
          subtractGis(plant.loc.coordinates[1], northWestPoints.lat),
        ]
      }
    };
  });
}

const metrics = Immutable.fromJS([{
  key: 'height',
  label: 'Height', // For InputCombo
  placeholder: 'Enter height of plant', // Input hint
  type: 'length',
}, {
  key: 'girth',
  label: 'Girth',
  placeholder: 'Enter girth of plant',
  type: 'length',
}, {
  key: 'harvestCount',
  label: 'Harvest Count',
  placeholder: 'Enter number of items harvested',
  type: 'count',
}, {
  key: 'harvestWeight',
  label: 'Harvest Weight',
  placeholder: 'Enter weight of harvest',
  type: 'weight',
}, {
  key: 'firstBlossum',
  label: 'First Blossum',
  placeholder: 'Check when first blossom is seen',
  type: 'toggle',
}, {
  key: 'lastBlossum',
  label: 'Last Blossum',
  placeholder: 'Check when last blossom is seen',
  type: 'toggle',
}, {
  key: 'firstBud',
  label: 'First Bud',
  placeholder: 'Check when first bud is seen',
  type: 'toggle',
}, {
  key: 'harvestStart',
  label: 'Harvest Start',
  placeholder: 'Check when harvest starts',
  type: 'toggle',
}, {
  key: 'harvestEnd',
  label: 'Harvest End',
  placeholder: 'Check when harvest ends',
  type: 'toggle',
}, {
  key: 'leafShedStart',
  label: 'Leaf Shed Start',
  placeholder: 'Check when leaf shed (abscission) starts',
  type: 'toggle',
}, {
  key: 'leafShedEnd',
  label: 'Leaf Shed End',
  placeholder: 'Check when leaf shed (abscission) ends',
  type: 'toggle',
},
]);

/**
 * Merges the values collected for a Note with values available
 * @param {string} key - the key of the object to get
 * @returns {Immutable} - the metaMetric for that key
 */
function metricsGetType(key) {
  return metrics.find(value => value.key === key);
}

module.exports = {
  dateToInt,
  filterPlants,
  filterSortPlants,
  getGeo,
  hasGeo,
  intToDate,
  intToMoment,
  intToString,
  makeLayoutUrl,
  makeMongoId,
  makePlantsUrl,
  makeSlug,
  metrics,
  metricsGetType,
  plantFromBody,
  rebaseLocations,
  sortPlants,
  transformErrors,
};

// TODO: Move this file to a /shared/ folder.
