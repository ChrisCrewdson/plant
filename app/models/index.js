const validatejs = require('validate.js');
const note = require('./note');
const plant = require('./plant');

// const utils = require('../libs/utils');

//  The validator receives the following arguments:
//     value - The value exactly how it looks in the attribute object.
//     options - The options for the validator. Guaranteed to not be null or undefined.
//     key - The attribute name.
//     attributes - The entire attributes object.
//     globalOptions - The options passed when calling validate
//                     (will always be an object, non null).
//
// If the validator passes simply return null or undefined. Otherwise return a string or an
// array of strings containing the error message(s).
// Make sure not to append the key name, this will be done automatically.


// Validate an integeter date. Should be in the range of
// something like 17000101 to 20201231. Not sure why we'd
// have dates beyond the current day...
validatejs.validators.intDateValidate = (value, options) => {
  if (Number.isNaN(value)) {
    // console.warn(`Date validate expected ${value} to not be NaN`);
    return `^${options.name} must be a valid date in the format MM/DD/YYYY`;
  }

  if (!value && !options.presence) {
    return null;
  }

  if (typeof value !== 'number') {
    // console.warn(`Date validate expected ${value} to be a number`);
    return `^${options.name} must be a number`;
  }

  // Don't allow dates before 1 Jan 1700
  if (value < 17000101) {
    // console.warn(`Date validation expected ${value} to be above 17000101`);
    return `^${options.name} must be after 1st Jan 1700`;
  }

  if (value.toString().length !== 8) {
    // console.error(`Date validation expected ${value} to have a string length of 8`);
    return `^${options.name} must be in format YYYYMMDD`;
  }

  const year = Math.round(value / 10000);
  if (year < 1700) {
    // console.warn(`Date validate expected ${value} to have a year greater than 1700`);
    return `^${options.name} must have a valid year, year found was ${year}`;
  }

  const month = Math.round(value / 100) - (year * 100);
  if (month < 1 || month > 12) {
    // console.warn(`Date validate expected ${value} to have a month from 01 to 12`);
    return `^${options.name} must have a valid month, value found was ${month}`;
  }

  const dayOfMonth = Math.round(value) - ((year * 10000) + (month * 100));
  if (dayOfMonth < 1 || dayOfMonth > 31) {
    // console.warn(`Date validate expected ${value} to have a day from 01 to 31`);
    return `^${options.name} must have a valid day-of-month, value found was ${dayOfMonth}`;
  }

  return null;
};

module.exports = {
  note, plant,
};

/*
Dates and transformations:

Inserting:
1. UI - date entered as string
2. onSave - date converted to number and validated
3. Ajax client put/post - date converted to string (implicit)
4. Ajax server put/post - date converted back to number and stored in DB as number

Getting:
1. From DB retrieved as number.
2. Ajax Server/Client - get - date retrieve as number
3. UI edit mode - date converted to string for editing

*/
