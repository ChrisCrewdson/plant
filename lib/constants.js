
const appConstants = require('../app/libs/constants');

module.exports = Object.freeze(Object.assign({}, appConstants, {
  sessionKey: 'plantSession', // Key for cookie to hold session. Managed by Passport
}));
