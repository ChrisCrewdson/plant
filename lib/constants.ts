
// Key for cookie to hold session. Managed by Passport
export const sessionKey = 'plantSession';
