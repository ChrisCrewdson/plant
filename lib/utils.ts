import { ImageSizeName } from './db/mongo/model-note';
import { UploadedNoteFile } from './types/model-note';

const firstDirectory = process.env.NODE_ENV === 'production' ? 'up' : 'test';

/**
 * The image is the object stored in the images array in the note.
 */
export const makeS3KeyFromImage = (image: UploadedNoteFile, size: ImageSizeName = 'orig',
): string => `${firstDirectory}/${size}/${image.id}${image.ext && '.'}${image.ext}`;
