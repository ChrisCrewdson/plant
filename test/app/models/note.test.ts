import _ from 'lodash';
import * as validators from '../../../app/models';
import * as constants from '../../../app/libs/constants';
import utils from '../../../app/libs/utils';

const { makeMongoId } = utils;
const { note: noteValidator } = validators;

describe('/app/models/note', () => {
  describe('basic validation', () => {
    test('should pass minimum validation', () => {
      const note: BizNote = {
        _id: '5c77545608a6cd11fdb0ca86',
        date: 20160101,
        plantIds: ['5c77545608a6cd11fdb0ca87'],
        note: 'some text',
        userId: makeMongoId(),
      };
      const noteCopy = _.cloneDeep(note);

      const transformed = noteValidator(note);
      expect(transformed).toMatchSnapshot();
      expect(noteCopy).toEqual(note);
    });

    test('should fail validation', () => {
    // All items in note should be invalid
      const note = {
        _id: '0e55d91cb33d42', // Not a MongoId
        date: 'Not a Number',
        plantIds: ['9ec5c8ffcf885bf'], // Not a MongoId in array
        note: {}, // not a string
      } as unknown as BizNote;

      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toMatchSnapshot();
        expect(noteCopy).toEqual(note);
      }
      expect.assertions(2);
    });

    test('should strip out props not in the schema', () => {
      const note = {
        _id: '5c2aefa2da47a52adc1c4651',
        date: 20160101,
        fakeName1: 'Common Name',
        fakeName2: 'Description',
        note: 'some text',
        plantId: 'fake plant id',
        plantIds: ['5c2aefa687358b2af50246d6'],
        userId: 'fake user id',
      } as BizNote;

      const noteCopy = _.cloneDeep(note);

      const transformed = noteValidator(note);

      expect(noteCopy).toEqual(note); // no mutation of original note

      expect(transformed).toMatchSnapshot();
    });

    test('should add _id if it is a new record', () => {
      const note: BizNote = {
        date: 20160101,
        plantIds: ['5c2aefa687358b2af50246d6'],
        note: 'some text',
      } as BizNote;
      const noteCopy = _.cloneDeep(note);

      const transformed = noteValidator(note);

      expect(noteCopy).toEqual(note);

      expect(transformed.plantIds).toEqual(note.plantIds);

      expect(transformed).toMatchSnapshot({
        _id: expect.stringMatching(constants.mongoIdRE),
      });
    });

    test('should fail if plantIds is empty', (done) => {
      const note: BizNote = {
        _id: '5c2aefa687358b2af50246d6',
        date: 20160101,
        note: 'some text',
        plantIds: [],
        userId: 'fake user id',
      };
      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toBeTruthy();
        expect(err.plantIds).toBe('You must select at least 1 plant for this note.');
        expect(noteCopy).toEqual(note);
        done();
      }
    });

    test('should fail if plantIds is missing', (done) => {
      const note = {
        _id: makeMongoId(),
        date: 20160101,
        note: 'some text',
      } as unknown as BizNote;
      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toBeTruthy();
        expect(err.plantIds).toBe('Plant ids is required');
        expect(noteCopy).toEqual(note);
        done();
      }
    });

    test('should fail if plantIds is not an array', (done) => {
      const note: BizNote = {
        _id: makeMongoId(),
        date: 20160101,
        note: 'some text',
        plantIds: makeMongoId(),
      } as unknown as BizNote;
      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toBeTruthy();
        expect(err.plantIds).toBe('Plant ids must be an array');
        expect(noteCopy).toEqual(note);
        done();
      }
    });
  });

  describe('metric validation', () => {

  });

  describe('note.model/images validation', () => {
    const image: NoteImage = {
      ext: 'jpg',
      id: '5c2af2b712d4132fa1e69d3a',
      originalname: 'apple tree',
      size: 123456,
      sizes: [],
    };

    test('should pass with an empty images array', () => {
      const note: BizNote = {
        _id: makeMongoId(),
        date: 20160101,
        images: [],
        note: 'some text',
        plantIds: [makeMongoId()],
        userId: 'fake user id',
      };
      const noteCopy = _.cloneDeep(note);


      const transformed = noteValidator(note);
      expect(Object.keys(transformed)).toHaveLength(5);
      expect(transformed._id).toBe(note._id);
      expect(transformed.note).toBe(note.note);
      expect(transformed.userId).toBeUndefined();
      expect(noteCopy).toEqual(note);
    });

    test('should pass with valid images', () => {
      const note: BizNote = {
        _id: makeMongoId(),
        date: 20160101,
        images: [image],
        note: 'some text',
        plantIds: [makeMongoId()],
        userId: 'fake user id',
      };
      const noteCopy = _.cloneDeep(note);

      const transformed = noteValidator(note);
      expect(Object.keys(transformed)).toHaveLength(5);
      expect(transformed._id).toBe(note._id);
      expect(transformed.note).toBe(note.note);
      expect(transformed.userId).toBeUndefined();
      expect(noteCopy).toEqual(note);
    });

    test('should pass with valid images and an empty note', () => {
      const note: BizNote = {
        _id: makeMongoId(),
        date: 20160101,
        images: [image],
        note: '',
        plantIds: [makeMongoId()],
        userId: 'fake user id',
      };
      const noteCopy = _.cloneDeep(note);

      const transformed = noteValidator(note);
      expect(Object.keys(transformed)).toHaveLength(5);
      expect(transformed._id).toBe(note._id);
      expect(transformed.note).toBe(note.note);
      expect(transformed.userId).toBeUndefined();
      expect(noteCopy).toEqual(note);
    });

    test('should pass with valid images and a missing note', () => {
      const note: BizNote = {
        _id: makeMongoId(),
        date: 20160101,
        images: [image],
        plantIds: [makeMongoId()],
        userId: 'fake user id',
      };
      const noteCopy = _.cloneDeep(note);

      const transformed = noteValidator(note);
      expect(Object.keys(transformed)).toHaveLength(4);
      expect(transformed._id).toBe(note._id);
      expect(transformed.userId).toBeUndefined();
      expect(noteCopy).toEqual(note);
    });

    test('should fail if images is not an array', (done) => {
      const note = {
        _id: makeMongoId(),
        date: 20160101,
        images: makeMongoId(),
        note: 'some text',
        plantIds: [makeMongoId()],
      } as unknown as BizNote;
      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toBeTruthy();
        expect(err.images).toBe('Images must be an array');
        expect(noteCopy).toEqual(note);
        done();
      }
    });

    test('should fail if images id is not a mongoId', (done) => {
      const note = {
        _id: makeMongoId(),
        date: 20160101,
        images: [{ ...image, id: 123 }],
        note: 'some text',
        plantIds: [makeMongoId()],
      } as unknown as BizNote;
      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toBeTruthy();
        expect(err.images).toBe('Images must be valid image objects');
        expect(noteCopy).toEqual(note);
        done();
      }
    });

    test('should fail if images ext is not a string', (done) => {
      const note: BizNote = {
        _id: makeMongoId(),
        date: 20160101,
        images: [{ ...image, id: '123' }],
        note: 'some text',
        plantIds: [makeMongoId()],
        userId: 'fake user id',
      };
      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toBeTruthy();
        expect(err.images).toBe('Images must be valid image objects');
        expect(noteCopy).toEqual(note);
        done();
      }
    });

    test('should fail if images originalname is not a string', (done) => {
      const note = {
        _id: makeMongoId(),
        date: 20160101,
        images: [{ ...image, originalname: 123 }],
        note: 'some text',
        plantIds: [makeMongoId()],
      } as unknown as BizNote;
      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toBeTruthy();
        expect(err.images).toBe('Images must be valid image objects');
        expect(noteCopy).toEqual(note);
        done();
      }
    });

    test('should convert image size if it is a string number', () => {
      const note = {
        _id: '5c2af2b712d4132fa1e69d39',
        date: 20160101,
        images: [{ ...image, size: '123' }],
        note: 'some text',
        plantIds: ['5c2af2b712d4132fa1e69d38'],
      } as unknown as BizNote;
      const noteCopy = _.cloneDeep(note);

      const transformed = noteValidator(note);

      expect(transformed).toMatchSnapshot();
      expect(noteCopy).toEqual(note);
    });

    test('should fail if images ext is longer than 20', (done) => {
      const note: BizNote = {
        _id: makeMongoId(),
        date: 20160101,
        images: [{ ...image, ext: '123456789012345678901' }],
        note: 'some text',
        plantIds: [makeMongoId()],
        userId: 'fake user id',
      };
      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toBeTruthy();
        expect(err.images).toBe('Images must be valid image objects');
        expect(noteCopy).toEqual(note);
        done();
      }
    });

    test('should fail if images has extra props', (done) => {
      const note: BizNote = {
        _id: makeMongoId(),
        date: 20160101,
        images: [{ ...image, extra: 'jpg' }],
        note: 'some text',
        plantIds: [makeMongoId()],
      } as unknown as BizNote;
      const noteCopy = _.cloneDeep(note);

      try {
        noteValidator(note);
      } catch (err) {
        expect(err).toMatchSnapshot();
        expect(noteCopy).toEqual(note);
        done();
      }
    });
  });
});
