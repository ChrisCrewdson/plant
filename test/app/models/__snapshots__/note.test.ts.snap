// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/app/models/note basic validation should add _id if it is a new record 1`] = `
Object {
  "_id": StringMatching /\\^\\[0-9a-f\\]\\{24\\}\\$/i,
  "date": 20160101,
  "note": "some text",
  "plantIds": Array [
    "5c2aefa687358b2af50246d6",
  ],
}
`;

exports[`/app/models/note basic validation should fail validation 1`] = `
Object {
  "_id": " id is invalid",
  "date": "Date must be a number",
  "note": "Note has an incorrect length",
  "plantIds": "Plant ids must be MongoIds",
}
`;

exports[`/app/models/note basic validation should pass minimum validation 1`] = `
Object {
  "_id": "5c77545608a6cd11fdb0ca86",
  "date": 20160101,
  "note": "some text",
  "plantIds": Array [
    "5c77545608a6cd11fdb0ca87",
  ],
}
`;

exports[`/app/models/note basic validation should strip out props not in the schema 1`] = `
Object {
  "_id": "5c2aefa2da47a52adc1c4651",
  "date": 20160101,
  "note": "some text",
  "plantIds": Array [
    "5c2aefa687358b2af50246d6",
  ],
}
`;

exports[`/app/models/note note.model/images validation should convert image size if it is a string number 1`] = `
Object {
  "_id": "5c2af2b712d4132fa1e69d39",
  "date": 20160101,
  "images": Array [
    Object {
      "ext": "jpg",
      "id": "5c2af2b712d4132fa1e69d3a",
      "originalname": "apple tree",
      "size": 123,
      "sizes": Array [],
    },
  ],
  "note": "some text",
  "plantIds": Array [
    "5c2af2b712d4132fa1e69d38",
  ],
}
`;

exports[`/app/models/note note.model/images validation should fail if images has extra props 1`] = `
Object {
  "images": "Images must only have the following allowed props: id,ext,originalname,size,sizes and in one of the items in the array it found these props as well: extra",
}
`;
